docs/proposals/scheduling/policy-preemption/README.md

```diff
@@ -0,0 +1,256 @@
+At some point, I want to change the scheduling strategy of one of the businesses under one of namespaces, so I create a pp to match it only.
+I hope that the propagationPolicy can match it that have been matched by clusterPropagationPolicy.
+
+### Risks and Mitigations
+
+This proposal maintains the backward compatibility, the system built with previous
+versions of Karmada can be seamlessly migrated to the new version.
+The previous configurations(yamls) could be applied to the new version of Karmada and
+without any behavior change.
+
+### Notes/Constraints/Caveats (Optional)
+
+The priority based on preemption do not include implicit priority when `.spec.priority` of policies are the same.
+
+For example, the following two policies do not preempt.
+
+```
+# labelselectormatch.yaml
+apiVersion: policy.karmada.io/v1alpha1
+kind: PropagationPolicy
+metadata:
+  name: propagation-labelselectormatch
+spec:
+  resourceSelectors:
+    - apiVersion: apps/v1
+      kind: Deployment
+      labelSelector:
+        matchLabels:
+          app: nginx
+  placement:
+    clusterAffinity:
+      clusterNames:
+        - member2
+---
+# namematch.yaml
+apiVersion: policy.karmada.io/v1alpha1
+kind: PropagationPolicy
+metadata:
+  name: propagation-namematch
+spec:
+  resourceSelectors:
+    - apiVersion: apps/v1
+      kind: Deployment
+      name: nginx
+  placement:
+    clusterAffinity:
+      clusterNames:
+        - member3
+```
+And when preemption is enabled, pp will preempt cpp regardless of the priority.
+
+Overall priority: high-priority pp > low-priority pp > high-priority cpp > low-priority cpp
+
+## Design Details
+
+### API change
+
+#### PropagationPolicy API change
+
+This proposal proposes a new field `Preemption` for determining the behavior for preempting.
+By default, preemption is disabled.
+
+```go
+type PropagationSpec struct {
+    // +kubebuilder:default=0
+    Priority *int `json:"priority,omitempty"`
+
+    // Preemption declares the behaviors for preempting.
+    // Valid options are Always and Never.
+    //
+    // +kubebuilder:default="Never"
+    // +kubebuilder:validation:Enum=Always;Never
+    // +optional
+    Preemption PreemptionBehavior `json:"preemption,omitempty"`
+}
+```
+
+```go
+// PreemptionBehavior describes whether and how to preempt resources that are
+// claimed by lower-priority PropagationPolicy(ClusterPropagationPolicy).
+// +enum
+type PreemptionBehavior string
+
+const (
+    // PreemptAlways means that preemption is allowed.
+    //
+    // If it is applied to a PropagationPolicy, it can preempt any resource as
+    // per Priority, regardless of whether it has been claimed by a PropagationPolicy
+    // or a ClusterPropagationPolicy, as long as it can match the rules defined
+    // in ResourceSelector. In addition, if a resource has already been claimed by a 
+    // ClusterPropagationPolicy, the PropagationPolicy can still preempt it
+    // without considering Priority.
+    //
```

Comment:
> // if a resource has already been claimed by a 
// ClusterPropagationPolicy, the PropagationPolicy can still preempt it
    // without considering Priority.


Do you mean if a low-priority  PropagationPolicy can preempt a high-priority ClusterPropagationPolicy?

Reply:
Yeah, this is consistent with the behavior of the first match(always look for pp first and then cpp).

Reply:
> Yeah, this is consistent with the behavior of the first match(always look for pp first and then cpp).


Yes, but there is a condition: this resource has not been claimed by any pp or cpp. If a resource has already been claimed by a cpp, then the pp will also not take effect.

According to the new proposal(`preemption: Always`), the cpp will become invalid regardless of whether this resource has been declared by a cpp or not, then the pp will take effect. Right?

Reply:
If this resource has been claimed by a cpp, a pp will preempt the resource if it enables preemption and also match it.
What I want to emphasize is that whether it's an initial match or a preemption, we always follow the order:
high-priority pp > low-priority pp > high-priority cpp > low-priority cpp.